File: src/game/systems/aggression.lua
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local Attack = require 'game.components.attack'
local Aggression = require 'game.components.aggression'
local Fixture = require 'game.components.fixture'
local Position = require 'game.components.position'
local Timer = require 'game.components.timer'
local collision = require 'game.utils.collision'

local aspect = Aspect:new({Aggression, Fixture, Position, Timer})
local AggressionSystem = System:new('aggression', aspect)

local STOP_ATTACK_DURATION = 3

function startAttack(origin, target)
  collision.entity(origin):add(Attack.new(1, collision.entity(target)))
end

function stopAttack(origin)
  local timer = collision.entity(origin):as(Timer)
  timer.timers.stopAttack = {duration = STOP_ATTACK_DURATION}
end

function AggressionSystem:update(dt)
  for _, entity in pairs(self.entities) do
    local aggression = entity:as(Aggression)
    local fixture = entity:as(Fixture)
    local timer = entity:as(Timer)

    if timer.timers.stopAttack and timer.timers.stopAttack.duration == 0 then
      timer.timers.stopAttack = nil
      entity:remove(Attack)
    end

    aggression.fixture:getBody():setPosition(
      fixture.fixture:getBody():getPosition()
    )
  end
end

function AggressionSystem:draw()
  for _, entity in pairs(self.entities) do
    local aggression = entity:as(Aggression)
    local body = aggression.fixture:getBody()
    local shape = aggression.fixture:getShape()
    local shapeType = shape:getType()

    love.graphics.setColor(255, 0, 0, 0.25)
    love.graphics.polygon('fill', body:getWorldPoints(shape:getPoints()))
    love.graphics.setColor(255, 255, 255, 1)
  end
end

function AggressionSystem:beginContact(a, b, contact)
  if collision.is('player', a) and collision.isAggression(b) then
    startAttack(b, a)
  elseif collision.isAggression(a) and collision.is('player', b) then
    startAttack(a, b)
  end
end

function AggressionSystem:endContact(a, b, contact)
  if collision.is('player', a) and collision.isAggression(b) then
    stopAttack(b, a)
  elseif collision.isAggression(a) and collision.is('player', b) then
    stopAttack(a, b)
  end
end

return AggressionSystem

File: client2/src/game/systems/Aggression.ts
import { Aspect } from '../../ecs/Aspect'
import { Entity } from '../../ecs/Entity'
import { System } from '../../ecs/System'
import { Aggression } from '../components/Aggression'
import { Attack } from '../components/Attack'
import { GameObject } from '../components/GameObject'
import { Player } from '../components/Player'
import { Position } from '../components/Position'
import { Waypoint } from '../components/Waypoint'
import { SystemFlag } from '../flags'
import { check, hasComponent } from '../utils/collision'
import { isPolygonShape } from '../utils/fixture'

const isAggression = (fixture: Fixture) => {
  return fixture.getUserData().isAggression
}

const setWaypointActive = (active: boolean, entity: Entity) => {
  const waypoint = entity.as(Waypoint)

  if (waypoint) {
    waypoint.active = active
  }
}

export class AggressionSystem extends System {
  static _id = 'Aggression'
  _id = AggressionSystem._id

  static _flag = SystemFlag.Aggression
  _flag = AggressionSystem._flag

  static _aspect = new Aspect([Aggression, GameObject, Position])
  _aspect = AggressionSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      const aggression = entity.as(Aggression)
      const gameObject = entity.as(GameObject)

      if (!aggression || !gameObject) {
        return
      }

      const position = gameObject.fixture.getBody().getPosition()

      aggression.fixture.getBody().setPosition(...position)
    })
  }

  beginContact = (a: Fixture, b: Fixture, contact: Contact) => {
    const result = check(a, b, [isAggression, hasComponent(Player)])

    if (!result) {
      return
    }

    const [attacker, target] = result

    const aggression = attacker.as(Aggression)

    if (!aggression) {
      return
    }

    aggression.clearDurationTimer()

    attacker.add(
      new Attack(aggression.followDistance, aggression.followVelocity, target)
    )

    setWaypointActive(false, attacker)
  }

  endContact = (a: Fixture, b: Fixture, contact: Contact) => {
    const result = check(a, b, [isAggression, hasComponent(Player)])

    if (!result) {
      return
    }

    const [attacker, target] = result

    const aggression = attacker.as(Aggression)

    if (!aggression) {
      return
    }

    aggression.setDurationTimer(() => {
      attacker.remove(Attack)

      setWaypointActive(true, attacker)
    })
  }

  draw = () => {
    this.entities.forEach(entity => {
      const aggression = entity.as(Aggression)

      if (aggression) {
        const body = aggression.fixture.getBody()
        const shape = aggression.fixture.getShape()

        if (isPolygonShape(shape)) {
          const points = body.getWorldPoints(...shape.getPoints())

          love.graphics.setColor(255, 0, 0, 0.25)
          love.graphics.polygon('fill', ...points)
          love.graphics.setColor(255, 255, 255, 1)
        }
      }
    })
  }
}
